src/rust/engine/process_execution/src/docker.rs

```diff
@@ -163,6 +163,52 @@ impl super::CommandRunner for CommandRunner {
   }
 }
 
+struct ContainerRemovalGuard {
+  docker: Docker,
+  container_id: Option<String>,
+  executor: Executor,
+}
+
+impl ContainerRemovalGuard {
+  pub fn new(docker: Docker, container_id: String, executor: Executor) -> Self {
+    Self {
+      docker,
+      container_id: Some(container_id),
+      executor,
+    }
+  }
+
+  pub fn preserve(&mut self) {
+    self.container_id = None;
+  }
+
+  pub async fn remove(&mut self) -> Result<(), String> {
+    if let Some(container_id) = self.container_id.take() {
+      let remove_options = bollard::container::RemoveContainerOptions {
+        force: true,
+        ..bollard::container::RemoveContainerOptions::default()
+      };
+
+      let _ = self
+        .docker
+        .remove_container(&container_id, Some(remove_options))
+        .await
+        .map_err(|err| format!("Failed to remove container `{}`: {:?}", &container_id, err))?;
+    }
+
+    Ok(())
+  }
+}
+
+impl Drop for ContainerRemovalGuard {
+  fn drop(&mut self) {
+    let executor = self.executor.clone();
+    if let Err(err) = executor.block_on(self.remove()) {
+      log::warn!("Error while removing container: {:?}", err);
+    }
```

Comment:
If you were to `tokio::spawn` here rather than `block_on`, you'd be able to drop asynchronously.

This will be worth doing due to cancellation, I think.